runtime/runtime.go

```diff
@@ -2145,3 +2184,207 @@ func NewBlockValue(block Block) interpreter.BlockValue {
+			if !ok {
+				panic("revoke method requires the first argument to be an integer")
+			}
+
+			var err error
+			var accountKey *AccountKey
+			wrapPanic(func() {
+				accountKey, err = runtimeInterface.RemoveAccountKey(addressValue.ToAddress(), index.ToInt())
+			})
+			if err != nil {
+				panic(err)
+			}
+
+			r.emitAccountEvent(
+				stdlib.AccountKeyAddedEventType,
+				runtimeInterface,
+				[]exportableValue{
+					newExportableValue(addressValue, nil),
+					newExportableValue(index, nil),
+				},
+			)
+
+			accountKeyValue := NewAccountKeyValue(accountKey)
+			return trampoline.Done{Result: accountKeyValue}
+		},
+	)
+}
+
+func (r *interpreterRuntime) newPublicAccountKeys(addressValue interpreter.AddressValue, runtimeInterface Interface) *interpreter.CompositeValue {
+	return interpreter.NewPublicAccountKeysValue(
+		r.newAccountKeysGetFunction(
+			addressValue,
+			runtimeInterface,
+		),
+	)
+}
+
+func NewPublicKeyFromValue(publicKey *interpreter.CompositeValue) *PublicKey {
+
+	// publicKey field
+	key, ok := publicKey.Fields.Get(sema.PublicKeyPublicKeyField)
+	if !ok {
+		panic("public key value is not set")
+	}
+
+	byteArray, err := interpreter.ByteArrayValueToByteSlice(key)
+	if err != nil {
+		panic(fmt.Errorf("public key needs to be a byte array. %w", err))
+	}
+
+	// sign algo field
+	signAlgoField, ok := publicKey.Fields.Get(sema.PublicKeySignAlgoField)
+	if !ok {
+		panic("sign algorithm is not set")
+	}
+
+	signAlgoValue, ok := signAlgoField.(*interpreter.CompositeValue)
+	if !ok || signAlgoValue.QualifiedIdentifier != sema.SignatureAlgorithmTypeName {
+		panic("sign algorithm needs to be of type `SignAlgorithm`")
+	}
+
+	rawValue, ok := signAlgoValue.Fields.Get(sema.EnumRawValueFieldName)
+	if !ok {
+		panic("cannot find sign algorithm raw value")
+	}
+
+	signAlgoRawValue, ok := rawValue.(interpreter.IntValue)
+	if !ok {
+		panic("enum raw value needs to be subtype of integer")
+	}
+
+	return &PublicKey{
+		PublicKey: byteArray,
+		SignAlgo:  SignatureAlgorithm(signAlgoRawValue.ToInt()),
+	}
+}
+
+func NewPublicKeyValue(publicKey *PublicKey) *interpreter.CompositeValue {
+	return interpreter.NewPublicKeyValue(
+		interpreter.ByteSliceToByteArrayValue(publicKey.PublicKey),
+		interpreter.NewNativeEnumCaseValue(sema.SignatureAlgorithmType, int(publicKey.SignAlgo)),
+	)
+}
+
+func NewAccountKeyValue(accountKey *AccountKey) *interpreter.CompositeValue {
+	return interpreter.NewAccountKeyValue(
+		interpreter.NewIntValueFromInt64(int64(accountKey.KeyIndex)),
+		NewPublicKeyValue(accountKey.PublicKey),
+		interpreter.NewNativeEnumCaseValue(sema.HashAlgorithmType, int(accountKey.HashAlgo)),
+		interpreter.NewUFix64ValueWithInteger(uint64(accountKey.Weight)),
+		interpreter.BoolValue(accountKey.IsRevoked),
+	)
+}
+
+func NewHashAlgorithmFromValue(value interpreter.Value) HashAlgorithm {
+	hashAlgoValue, ok := value.(*interpreter.CompositeValue)
+	if !ok || hashAlgoValue.QualifiedIdentifier != sema.HashAlgorithmTypeName {
```

Comment:
Maybe also assert location is nil

Reply:
Similar to https://github.com/onflow/cadence/pull/633#discussion_r592774043, these should have been validated at the time of importing, and all the other cases should have been statically validated. So removed these validations from here as well.